docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb

```diff
@@ -0,0 +1,681 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "'No AVX2, quantization model might be slow'"
+      ]
+     },
+     "execution_count": 10,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "subprocess.run([sys.executable, '-m', 'pip', 'install', '--user', 'cpufeature'])\n",
+    "import cpufeature\n",
+    "cpufeature.CPUFeature['AVX2'] or 'No AVX2, quantization model might be slow'"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can use onnxruntime.nuphar.model_quantizer to quantize floating point GEMM/GEMVs. Assuming GEMM/GEMV takes form of input * weights, weights are statically quantized per-column, and inputs are dynamically quantized per-row."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "from onnxruntime.nuphar.model_quantizer import convert_matmul_model\n",
+    "quantized_model = 'Scan_LSTMx4_int8.onnx'\n",
+    "convert_matmul_model(scan_model, quantized_model)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now run the quantized model, and check accuracy. Please note that quantization brings some error, so we relaxed on compare threshold."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "sess_quantized = onnxruntime.InferenceSession(quantized_model)\n",
+    "output_quantized = sess_quantized.run([], feed)\n",
+    "assert np.allclose(output[0], output_quantized[0], rtol=1e-3, atol=1e-3)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now check quantized model performance:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "'quantized: 1.1350197000000009 seconds, non-quantized: 2.5135028000000013 seconds'"
+      ]
+     },
+     "execution_count": 13,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "start_quantized = timer()\n",
+    "for i in range(repeats):\n",
+    "    output = sess_quantized.run([], feed)\n",
+    "end_quantized = timer()\n",
+    "\n",
+    "'quantized: {} seconds, non-quantized: {} seconds'.format(end_quantized - start_quantized, end_nuphar - start_nuphar)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 4. Working on a real model: Bidirectional Attention Flow (BiDAF)\n",
+    "BiDAF is a machine comprehension model that uses LSTM. The inputs to this model are paragraphs of context and query, and the outputs are start/end index of words in the context that answers the query.\n",
```

Comment:
typo: 
"of context and query, and the outputs are start/end index of words in the context that answers the query" ->

"of contexts and queries, and the outputs are start/end indexes of the words in the contexts that answer the queries"